### ISSUE TITLE
Failed to start gauge API: Timed out connecting to java (1.1.1)

### ISSUE THREAD

**Issue by haroon-sheikh** [getgauge/gauge-java#455]
### Describe the bug
Getting loads of these error when running on the latest version of gauge and gauge-java

### What command(s) did you run when you found the bug?

For e.g. 
```
mvn clean test
```

### Output, stack trace or logs related to the bug 

```sh
Failed to start gauge API: Timed out connecting to java
```

#### Gauge (Output of `gauge -v`)

```sh
Gauge version: 1.1.1
Plugins
-------
html-report (4.0.11)
java (0.7.8)
reportportal (1.6.0)
reportserver (0.2.0)
screenshot (0.0.1)
xml-report (0.2.3)
```

#### Operating System information
OSx

**Comment by BugDiver** [comment]
TLDR;
The gauge-java `launcher` (shell/pwsh scripts) having difficulties while computing the classpath, causing ☝️ issues.

Earlier we fixed an issue to support external libraries (jar files in the lib). To do so gauge-java start populating the `step-registry` via reflection along with static parsing of existing source files. Now in order to get external libs reflected it needs to in the classpath. The classpath is managed via launcher scripts or maven/Gradle plugin. The execution worked fine cause when running the test via maven/Gradle (`mvn gauge: execute` or `./gradlew gauge`) and the classpath was managed.
But when working with IDE (VS code or any other which uses `gauge lsp`), the IDEs were invoking gauge directly (without using maven/Gradle) hence the runner launched via `IDE -> gauge -> runner` did not get the expected classpath (external libs). In order to tackle this, we added a task `classpath` in maven/Gradle plugin which gives the classpath and the launcher scripts can set it.

Now seems like managing all these can be really error-prone since depending on the type of dependencies there can be errors/warning which can mess up with the task output.


As per discussion with the @getgauge/core here is what we are gonna do,

* Make launcher script as simple as possible (remove computing classpath for maven/Gradle)
*. Remove `classpath` tasks from plugins and it's usages.
* To make IDE's work, add a mojo/task in respective plugins which will work like `execute` (set the classpath and run gauge with respective subcommand)
* IDE's will invoke gauge via the plugin for LSP.

**Comment by BugDiver** [comment]
@haroon-sheikh As discussed with the @getgauge/core We have decided to bring back the golang launcher (It was there prior 0.7.4, ref https://github.com/getgauge/gauge-java/issues/324). 

After 0.7.4 we have faced a lot of issues related to the launcher scripts, due to the limitation of shell/pwsh. And the launcher has evolved a bit as well (version check etc), which is quite error-prone in sh/pwsh.

I have pushed these changes in a branch called `fix-launcher`. The classpath computation part has been removed from the launcher. It needs to be handled as discussed in the previous comment (via vscode and mvn/gradle plugin). 

I will let you know once the test artifacts are available.

**Comment by BugDiver** [comment]
@haroon-sheikh The test artifacts are available (https://github.com/getgauge/gauge-java/suites/806799454/artifacts/8766219).  Please download and install the platform-specific plugin. Due to limitations of Github Actions I could not upload the artifacts as separate files so You have to download the whole folder and use the plugin suitable to your platform (Apologies).

**Comment by haroon-sheikh** [comment]
Thanks @BugDiver. I've now installed this, no issues so far with the initial runs. Will let you know if there's any.

**Comment by BugDiver** [comment]
@cyberjaime45 Forgot to loop you in. Could you also verify if it's working for you?

**Comment by haroon-sheikh** [comment]
Hi @BugDiver, Unfortunately, I'm still seeing the errors with the fix build:

```sh

17-06-2020 19:49:40.863 [Gauge] [ERROR] Failed to start runner. Timed out connecting to java
17-06-2020 19:49:40.864 [Gauge] [DEBUG] Skipping 261 specifications
17-06-2020 19:49:40.864 [Gauge] [ERROR] Failed to start runner. Timed out connecting to java
17-06-2020 19:49:40.865 [Gauge] [DEBUG] Skipping 261 specifications
17-06-2020 19:49:40.865 [Gauge] [ERROR] Failed to start runner. Timed out connecting to java
17-06-2020 19:49:40.865 [Gauge] [DEBUG] Skipping 261 specifications
17-06-2020 19:49:40.869 [Gauge] [ERROR] Failed to start runner. Timed out connecting to java
17-06-2020 19:49:40.869 [Gauge] [DEBUG] Skipping 261 specifications
17-06-2020 19:49:40.873 [Gauge] [ERROR] Failed to start runner. Timed out connecting to java
17-06-2020 19:49:40.873 [Gauge] [DEBUG] Skipping 261 specifications
17-06-2020 19:49:40.875 [Gauge] [ERROR] Failed to start runner. Timed out connecting to java
17-06-2020 19:49:40.875 [Gauge] [DEBUG] Skipping 261 specifications
17-06-2020 19:49:40.881 [Gauge] [ERROR] Failed to start runner. Timed out connecting to java
17-06-2020 19:49:40.881 [Gauge] [DEBUG] Skipping 261 specifications
17-06-2020 19:49:40.888 [Gauge] [ERROR] Failed to start runner. Timed out connecting to java
17-06-2020 19:49:40.888 [Gauge] [DEBUG] Skipping 261 specifications
17-06-2020 19:49:40.889 [Gauge] [ERROR] Failed to start runner. Timed out connecting to java
17-06-2020 19:49:40.890 [Gauge] [ERROR] Failed to start runner. Timed out connecting to java
17-06-2020 19:49:40.890 [Gauge] [ERROR] Failed to start runner. Timed out connecting to java
17-06-2020 19:49:40.890 [Gauge] [ERROR] Failed to start runner. Timed out connecting to java
17-06-2020 19:49:40.890 [Gauge] [ERROR] Failed to start runner. Timed out connecting to java
17-06-2020 19:49:40.890 [Gauge] [ERROR] Failed to start runner. Timed out connecting to java
17-06-2020 19:49:40.890 [Gauge] [ERROR] Failed to start runner. Timed out connecting to java
17-06-2020 19:49:40.890 [Gauge] [ERROR] Failed to start runner. Timed out connecting to java
17-06-2020 19:49:40.911 [html-report] [DEBUG] Transformed SuiteResult to report structure
17-06-2020 19:49:40.936 [html-report] [CRITICAL] write /.../html-report/index.html: file already closed
17-06-2020 19:49:40.940 [Gauge] [ERROR] Error occurred while waiting for plugin process to finish.
Error : exit status 1
17-06-2020 19:49:40.940 [Gauge] [ERROR] Unable to connect to plugin Html Report 4.0.11. rpc error: code = Unavailable desc = transport is closing
17-06-2020 19:49:40.940 [Gauge] [DEBUG] Killing Plugin Html Report 4.0.11
17-06-2020 19:49:40.940 [Gauge] [ERROR] Failed to kill plugin Html Report 4.0.11. os: process already finished
17-06-2020 19:49:40.966 [xml-report] [INFO] Successfully generated xml-report to => /.../reports/xml-report
17-06-2020 19:49:41.240 [Gauge] [DEBUG] Sending kill message to Xml Report plugin.
```

**Comment by haroon-sheikh** [comment]
@BugDiver I've noticed this happens more often when `enable_multithreading` is set to false. But seeing issues with it when set to true also.

**Comment by BugDiver** [comment]
@haroon-sheikh That's weird. I tried replicating it but worked well for me.
From the logs, it looks like the gauge is not able to start/connect with the plugins. 
Could you please follow these steps to check if the plugins are able to start?
Run all the steps inside the gauge project.

* Collect mvn classpath in file (Run mvn dependency:build-classpath -Dmdep.outputFile=cp.txt`
* `gauge_custom_classpath=${cat cp.txt} GAUGE_PROJECT_ROOT=${pwd} $HOME/.gauge/plugins/java/0.7.9/gauge-java --start`

**Comment by haroon-sheikh** [comment]
@BugDiver I'm seeing the same error when running this `gauge_custom_classpath=$(cat cp.txt) GAUGE_PROJECT_ROOT=$(pwd) $HOME/.gauge/plugins/java/0.7.9/gauge-java --start`

```sh
{"logLevel":"debug","message":"failed to get full path for mvn. lstat /Library/Java/JavaVirtualMachines/jdk-11.0.2.jdk/Contents/Home/:/Users/user/Library/Android/sdk/tools:/Users/user/Library/Android/sdk/platform-tools:/Users/user/.yarn/bin:/Users/user/.config/yarn/global/node_modules/.bin:/Users/user/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/go/bin:/Library/Apple/usr/bin:/Library/Frameworks/Mono.framework/Versions/Current/Commands:/Applications/Wireshark.app/Contents/MacOS: no such file or directory"}
{"logLevel":"debug","message":"failed to get full path for gauge. lstat /Library/Java/JavaVirtualMachines/jdk-11.0.2.jdk/Contents/Home/:/Users/user/Library/Android/sdk/tools:/Users/user/Library/Android/sdk/platform-tools:/Users/user/.yarn/bin:/Users/user/.config/yarn/global/node_modules/.bin:/Users/user/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/go/bin:/Library/Apple/usr/bin:/Library/Frameworks/Mono.framework/Versions/Current/Commands:/Applications/Wireshark.app/Contents/MacOS: no such file or directory"}
{"logLevel":"error","message":"Installed version of gauge-java (0.7.9) does not match with dependency gauge-java (0.7.8) specified in pom.xml file"}
```

**Comment by BugDiver** [comment]
@haroon-sheikh The errors which you are seeing is
* The `failed to get full path for` is because the launcher tries to find the absolute path using the PATH env, if it fails it uses the literal word (gauge,mvn etc). I do not think thats the problem,
* `Installed version of gauge-java (0.7.9)` is because the launcher and mvn dep has diff version Which is also not an huge issue.

Do you see anything after these errors, If the runner is able to start properly you should see something lke `Listening on port: SOME_PORT`
or you should see some other errors

**Comment by haroon-sheikh** [comment]
Running `gauge_custom_classpath=$(cat cp.txt) GAUGE_PROJECT_ROOT=$(pwd) $HOME/.gauge/plugins/java/0.7.9/gauge-java --start` shows `Listening on port: SOME_PORT`

However, running the actual tests `gauge run specs/` is just throwing `Failed to start gauge API: Timed out connecting to java`.

**Comment by BugDiver** [comment]
@haroon-sheikh Unfortunately I'm not able to replicate it on my end. Could you please create and a sample project where you are able to replicate it. You can also dockerize that project and share the image.

**Comment by haroon-sheikh** [comment]
@BugDiver Unfortunately, I'm not able to share the project. And not able to work out, what's different in this project that could cause this so that I'm able to create a sample project.

Is there anything else that I can do to provide further debugging?

**Comment by BugDiver** [comment]
> I'm not able to share the project.

We understand this and that's the reason I'm asking a sample project, which needs not a full-blown project.

> what's different in this project that could cause this so

We can't say if it's a project related thing. I'm asking for a project (where you are able to replicate the same error) so that we can figure it out. Best if you can dockerize it (You should be able to get the error inside the docker container), that way we can eliminate machine specifics

**Comment by haroon-sheikh** [comment]
@BugDiver Can we jump on a call maybe?

**Comment by BugDiver** [comment]
@haroon-sheikh Sure. We can discuss the timing and medium personal spectrum chat.

**Comment by madamore** [comment]
Trying to debug, running single spec:
Openjdk 11
intellij (2020.1.2 Community) + intellij gauge plugin 0.3.21
gauge 1.1.1 + java plugin 0.7.8 + maven gauge-java 0.7.8 = Failed to start gauge API: Timed out connecting to java
gauge 1.1.1 + java plugin 0.7.7 + maven gauge-java 0.7.7 = Works properly

**Comment by haroon-sheikh** [comment]
@madamore can you try setting `enable_multithreading=false` and see if that fixes for you.

**Comment by madamore** [comment]
> enable_multithreading

It was already set as "enable_multithreading=false"

**Comment by haroon-sheikh** [comment]
Have you tried changing to true if that makes a difference?

**Comment by madhavabk** [comment]
I have gauge 1.1.1 using IntelliJ editor.
It fails with message 
This version of gauge-java plugin does not support Java versions < 1.9
Please upgrade your Java version or use a version of gauge-java <= v0.7.4

Now i installed java 1.9 and put path in java.properties file
After this it will fail with message use gauge-maven-plugin 1.4.3 or greater

Now i change pom to use 1.4.3 instead my earlier 1.4.1
Failed to start gauge API: Timed out connecting to java

**Comment by sriv** [comment]
@madhavabk - do you see anything in the logs? inspect your logs folder and see if there are errors.

Also please ensure that once you upgrade gauge-java plugin, your pom/gradle references are also up-to-date.

**Comment by sriv** [comment]
An update to those who are following this issue - @haroon-sheikh and I have been troubleshooting offline, and we seem to have found a cause. Will post more when there's a fix/workaround.

**Comment by sriv** [comment]
Hi, we just released gauge-java `0.7.9`. It should be available in maven-central in a while.

This release has some improvements in the way the java runner is launched. It also adds more logs when run with log-level=debug.

If you still face this issue of `Failed to start gauge API: Timed out connecting to java` then, please run with `--log-level=debug` and:

- inspect your logs directory for `gauge.log` and see if there are any errors reported.
- the logs directory should contain `<pid>.out` files, which contain

**Comment by haroon-sheikh** [comment]
@sriv @NivedhaSenthil `gauge install java` is installing 0.7.8 and `gauge install java --version=0.7.9` is not found. Do you know how long it takes before it's available?

**Comment by cyberjaime45** [comment]
To be available in https://mvnrepository.com, it usually takes about 24 hours. 
You can also install manually by downloading the file and running 
`gauge install java --file  name-of-file.zip`

But you'll still need to wait  to be available in mvnrepository.com since the version also goes in your pom file (maven project)

**Comment by zabil** [comment]
Actually for plugin metadata it uses a cdn(jsdeliver). This is usually updated every 12-24 hours.
This is redirect used for plugin downloads
https://github.com/getgauge/downloads.gauge.org/blob/master/netlify.toml

**Comment by jonathangiber** [comment]
Thanks for the hard work, I'm running the last version without any crash now

**Comment by haroon-sheikh** [comment]
@sriv I'm still seeing the issue when multi_threading=false :(

```
Gauge version: 1.1.1
Plugins
-------
flash (0.0.2)
html-report (4.0.11)
java (0.7.9)
json-report (0.3.3)
reportportal (1.7.0)
reportserver (0.2.0)
screenshot (0.0.1)
spectacle (0.1.4)
xml-report (0.2.3)
```

**Comment by haroon-sheikh** [comment]
And seeing the following error when enable_multithreading=true

```sh
Error ----------------------------------

[java]
Error in thread 20
null
java.lang.NullPointerException
        at com.thoughtworks.gauge.registry.StepRegistry.addStepImplementation(StepRegistry.java:35)
        at com.thoughtworks.gauge.scan.StepsScanner.buildStepRegistry(StepsScanner.java:46)
        at com.thoughtworks.gauge.scan.StepsScanner.scan(StepsScanner.java:33)
        at com.thoughtworks.gauge.scan.ClasspathScanner.scan(ClasspathScanner.java:35)
        at com.thoughtworks.gauge.connection.MessageProcessorFactory.getProcessor(MessageProcessorFactory.java:61)
        at com.thoughtworks.gauge.RunnerServiceHandler.lambda$initializeSpecDataStore$2(RunnerServiceHandler.java:76)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)

[java]
Error in thread 22
null
java.lang.NullPointerException
        at com.thoughtworks.gauge.registry.StepRegistry.addStepImplementation(StepRegistry.java:35)
        at com.thoughtworks.gauge.scan.StepsScanner.buildStepRegistry(StepsScanner.java:46)
        at com.thoughtworks.gauge.scan.StepsScanner.scan(StepsScanner.java:33)
        at com.thoughtworks.gauge.scan.ClasspathScanner.scan(ClasspathScanner.java:35)
        at com.thoughtworks.gauge.connection.MessageProcessorFactory.getProcessor(MessageProcessorFactory.java:61)
        at com.thoughtworks.gauge.RunnerServiceHandler.lambda$initializeSpecDataStore$2(RunnerServiceHandler.java:76)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)

[Gauge]
java runner quit unexpectedly. Check logs for more details.
ErrorMessage: transport is closing

[Gauge]
java runner quit unexpectedly. Check logs for more details.
ErrorMessage: transport is closing

Get Support ----------------------------
        Docs:          https://docs.gauge.org
        Bugs:          https://github.com/getgauge/gauge/issues
        Chat:          https://spectrum.chat/gauge

Your Environment Information -----------
        darwin, 1.1.1
        flash (0.0.2), html-report (4.0.11), java (0.7.9), json-report (0.3.3), reportportal (1.7.0), reportserver (0.2.0), screenshot (0.0.1), spectacle (0.1.4), xml-report (0.2.3)

```

**Comment by sriv** [comment]
@haroon-sheikh bummer! Anything relevant in the logs for enable_multithreading=false? 

And the error on multithreaded run is new. Will need to look deeper

**Comment by haroon-sheikh** [comment]
> @haroon-sheikh bummer! Anything relevant in the logs for enable_multithreading=false?
> 
> And the error on multithreaded run is new. Will need to look deeper

Nothing new, I still see `{"logLevel":"info","message":"Listening on port:12345"}` in each java.out files.

**Comment by haroon-sheikh** [comment]
And this in middle of the run:

```sh
Error in thread 15
null
java.lang.NullPointerException
        at com.thoughtworks.gauge.processor.ExecuteStepProcessor.process(ExecuteStepProcessor.java:35)
        at com.thoughtworks.gauge.RunnerServiceHandler.lambda$executeStep$7(RunnerServiceHandler.java:166)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)

[Gauge]
java runner quit unexpectedly. Check logs for more details.
ErrorMessage: transport is closing

[Gauge]
java runner quit unexpectedly. Check logs for more details.
ErrorMessage: transport is closing

Get Support ----------------------------
        Docs:          https://docs.gauge.org
        Bugs:          https://github.com/getgauge/gauge/issues
        Chat:          https://spectrum.chat/gauge

Your Environment Information -----------
        darwin, 1.1.1
        flash (0.0.2), html-report (4.0.11), java (0.7.9), json-report (0.3.3), reportportal (1.7.0), reportserver (0.2.0), screenshot (0.0.1), spectacle (0.1.4), xml-report (0.2.3)
Error ----------------------------------

[java]
Error in thread 15
null
java.lang.NullPointerException
        at com.thoughtworks.gauge.processor.ExecuteStepProcessor.process(ExecuteStepProcessor.java:35)
        at com.thoughtworks.gauge.RunnerServiceHandler.lambda$executeStep$7(RunnerServiceHandler.java:166)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)

[Gauge]
java runner quit unexpectedly. Check logs for more details.
ErrorMessage: transport is closing

Get Support ----------------------------
        Docs:          https://docs.gauge.org
        Bugs:          https://github.com/getgauge/gauge/issues
        Chat:          https://spectrum.chat/gauge

Your Environment Information -----------
        darwin, 1.1.1
        flash (0.0.2), html-report (4.0.11), java (0.7.9), json-report (0.3.3), reportportal (1.7.0), reportserver (0.2.0), screenshot (0.0.1), spectacle (0.1.4), xml-report (0.2.3)
```

**Comment by cyberjaime45** [comment]
**This issue still happening for:** @sriv 
 
Gauge version: 1.1.1
Commit Hash: 6e0d83f

Plugins:
flash (0.0.2)
html-report (4.0.12)
java (0.7.9)
json-report (0.3.3)
screenshot (0.0.1)

Note: If downgrade to Gauge v1.0.8 with Java v0.7.4. It works a lot better, a least don't have issues running tests. Any enhancements added to newer versions, the framework starts fail horrible. I hope you guys can figure it out because this issue is going no where.


**Jenkins console logs:**
Failed to start runner. Timed out connecting to java
Failed to start runner. Timed out connecting to java
Failed to start runner. Timed out connecting to java
Failed to start runner. Timed out connecting to java
Failed to start runner. Timed out connecting to java
Failed to start runner. Timed out connecting to java
Failed to start runner. Timed out connecting to java
I have to abort the Jenkins job otherwise it will running for ever stuck at "Failed to start runner. Timed out connecting to java"
In addition, server memory used jumps to 9.36 Gigabytes

`22-07-2020 21:28:36.938 [java] [DEBUG] stdout/stderr is logged to logs/staging/13192-java.out
22-07-2020 21:28:36.987 [java] [DEBUG] multithreading is set to false
22-07-2020 21:28:37.090 [java] [DEBUG] multithreading is set to false
22-07-2020 21:28:37.383 [java] [DEBUG] multithreading is set to false
22-07-2020 21:28:37.430 [java] [DEBUG] multithreading is set to false
22-07-2020 21:28:37.667 [java] [DEBUG] multithreading is set to false
22-07-2020 21:28:37.700 [java] [DEBUG] multithreading is set to false
22-07-2020 21:28:37.824 [java] [DEBUG] multithreading is set to false
22-07-2020 21:28:37.970 [java] [DEBUG] multithreading is set to false
22-07-2020 21:28:37.981 [java] [DEBUG] multithreading is set to false
22-07-2020 21:28:38.144 [java] [DEBUG] multithreading is set to false
22-07-2020 21:28:38.272 [java] [DEBUG] multithreading is set to false
22-07-2020 21:28:38.439 [java] [DEBUG] multithreading is set to false
22-07-2020 21:29:02.148 [java] [DEBUG] Using reflection to scan dependencies for gauge implementations...
22-07-2020 21:29:03.629 [java] [DEBUG] Using reflection to scan dependencies for gauge implementations...
22-07-2020 21:29:03.758 [java] [DEBUG] Using reflection to scan dependencies for gauge implementations...
22-07-2020 21:29:03.977 [java] [DEBUG] Using reflection to scan dependencies for gauge implementations...
22-07-2020 21:29:05.050 [java] [DEBUG] Using reflection to scan dependencies for gauge implementations...
22-07-2020 21:29:05.713 [java] [DEBUG] Using reflection to scan dependencies for gauge implementations...
22-07-2020 21:29:05.746 [java] [DEBUG] Using reflection to scan dependencies for gauge implementations...
22-07-2020 21:29:06.233 [java] [DEBUG] Using reflection to scan dependencies for gauge implementations...
22-07-2020 21:29:06.740 [java] [DEBUG] Using reflection to scan dependencies for gauge implementations...
22-07-2020 21:29:06.923 [java] [DEBUG] Using reflection to scan dependencies for gauge implementations...
22-07-2020 21:29:07.566 [java] [DEBUG] Using reflection to scan dependencies for gauge implementations...
22-07-2020 21:29:07.601 [java] [DEBUG] Using reflection to scan dependencies for gauge implementations...
22-07-2020 21:29:08.050 [java] [DEBUG] Using reflection to scan dependencies for gauge implementations...
22-07-2020 21:29:08.221 [java] [DEBUG] Using reflection to scan dependencies for gauge implementations...
22-07-2020 21:29:12.829 [java] [DEBUG] gauge-java took 36147milliseconds to load and scan. This should be less than 'runner_connection_timeout' config value.
22-07-2020 21:29:12.829 [java] [DEBUG] run 'gauge config runner_connection_timeout' and verify that it is < 36147
22-07-2020 21:29:12.829 [java] [DEBUG] starting gRPC server...
22-07-2020 21:29:13.780 [java] [DEBUG] gauge-java took 35954milliseconds to load and scan. This should be less than 'runner_connection_timeout' config value.
22-07-2020 21:29:13.780 [java] [DEBUG] run 'gauge config runner_connection_timeout' and verify that it is < 35954
22-07-2020 21:29:13.780 [java] [DEBUG] starting gRPC server...
22-07-2020 21:29:15.196 [java] [DEBUG] started gRPC server on port 37113
22-07-2020 21:29:15.357 [java] [DEBUG] gauge-java took 38326milliseconds to load and scan. This should be less than 'runner_connection_timeout' config value.
22-07-2020 21:29:15.357 [java] [DEBUG] run 'gauge config runner_connection_timeout' and verify that it is < 38326
22-07-2020 21:29:15.357 [java] [DEBUG] starting gRPC server..`
**and ........ (same file)**
`22-07-2020 21:29:18.391 [java] [DEBUG] starting gRPC server...
22-07-2020 21:29:18.458 [java] [DEBUG] started gRPC server on port 36699
22-07-2020 21:29:18.644 [java] [DEBUG] gauge-java took 40970milliseconds to load and scan. This should be less than 'runner_connection_timeout' config value.
22-07-2020 21:29:18.644 [java] [DEBUG] run 'gauge config runner_connection_timeout' and verify that it is < 40970
22-07-2020 21:29:18.644 [java] [DEBUG] starting gRPC server...
22-07-2020 21:29:18.770 [java] [DEBUG] started gRPC server on port 44355
22-07-2020 21:29:19.412 [java] [DEBUG] started gRPC server on port 40505
22-07-2020 21:29:19.578 [java] [DEBUG] started gRPC server on port 33323
22-07-2020 21:29:19.594 [java] [DEBUG] started gRPC server on port 46385
22-07-2020 21:29:19.833 [java] [DEBUG] gauge-java took 41687milliseconds to load and scan. This should be less than 'runner_connection_timeout' config value.
22-07-2020 21:29:19.833 [java] [DEBUG] run 'gauge config runner_connection_timeout' and verify that it is < 41687
22-07-2020 21:29:19.833 [java] [DEBUG] starting gRPC server...
22-07-2020 21:49:44.873 [java] [CRITICAL] process /usr/lib/jvm/jdk-11.0.7/bin/java with pid 12999 quit unexpectedly. exit status 143
22-07-2020 21:49:44.873 [Gauge] [ERROR] Error occurred while waiting for runner process to finish.
Error : exit status 1
22-07-2020 21:49:45.141 [Gauge] [ERROR] Error occurred while waiting for plugin process to finish.
Error : signal: terminated
22-07-2020 21:49:45.152 [Gauge] [ERROR] Error occurred while waiting for plugin process to finish.
Error : signal: terminated
22-07-2020 21:49:45.161 [Gauge] [ERROR] Error occurred while waiting for plugin process to finish.
Error : signal: terminated
22-07-2020 21:49:50.605 [java] [CRITICAL] process /usr/lib/jvm/jdk-11.0.7/bin/java with pid 14054 quit unexpectedly. exit status 143
22-07-2020 21:49:50.606 [Gauge] [ERROR] Error occurred while waiting for runner process to finish.
Error : exit status 1
22-07-2020 21:49:50.645 [java] [CRITICAL] process /usr/lib/jvm/jdk-11.0.7/bin/java with pid 13860 quit unexpectedly. exit status 143
22-07-2020 21:49:50.646 [Gauge] [ERROR] Error occurred while waiting for runner process to finish.`


**File: 1307-java.out (Partially added)**
`{"logLevel":"debug","message":"multithreading is set to false"}
{"logLevel":"debug","message":"Using reflection to scan dependencies for gauge implementations..."}
{"logLevel":"debug","message":"gauge-java took 40115milliseconds to load and scan. This should be less than 'runner_connection_timeout' config value."}
{"logLevel":"debug","message":"run 'gauge config runner_connection_timeout' and verify that it is < 40115"}
{"logLevel":"debug","message":"starting gRPC server..."}
{"logLevel":"debug","message":"started gRPC server on port 33323"}
{"logLevel":"info","message":"Listening on port:33323"}
{"logLevel":"debug","message":"Scanning packages for steps"}`

**Comment by sriv** [comment]
@cyberjaime45 - what is the output for `gauge config runner_connection_timeout` ? (If you are using the default, then it would be 30 seconds, but it seems that in your project gauge-java requires more than 40 seconds to scan, so you could try increasing the timeout to 60 seconds)

> I hope you guys can figure it out because this issue is going no where.
Trust me, we aren't very please about this either! It will be a lot faster and easier if we had a way to replicate the issue. 

>  In addition, server memory used jumps to 9.36 Gigabytes
Your project isn't using multithreading, and I see at least 14 streams (which means 14 jvm processes) - you could try setting `enable_multithreading=true` and see if that help.

**Comment by cyberjaime45** [comment]
@sriv 
This are my values 
runner_connection_timeout=200000
runner_request_timeout=500000
enable_multithreading = false

Should I change it? If so, what values you recommend ?

**Comment by sriv** [comment]
what happens if you set `enable_multithreading=true` ? 

Also just want to make sure that the below

```
runner_connection_timeout=200000
runner_request_timeout=500000
``` 

are set in gauge's global configuration, and not in project's `env`. You can verify by running `gauge config --list`

**Comment by jonathangiber** [comment]
I noticed the following error in one of my executions but overall is working fine, these cause many runners not to start at all in a 48 parallel execution:

[58.082s][warning][os,thread] Failed to start thread - pthread_create failed (EAGAIN) for attributes: stacksize: 1024k, guardsize: 0k, detached.
[58.269s][warning][os,thread] Failed to start thread - pthread_create failed (EAGAIN) for attributes: stacksize: 1024k, guardsize: 0k, detached.
[58.270s][warning][os,thread] Failed to start thread - pthread_create failed (EAGAIN) for attributes: stacksize: 1024k, guardsize: 0k, detached.
[58.262s][warning][os,thread] Failed to start thread - pthread_create failed (EAGAIN) for attributes: stacksize: 1024k, guardsize: 0k, detached.
[58.345s][warning][os,thread] Failed to start thread - pthread_create failed (EAGAIN) for attributes: stacksize: 1024k, guardsize: 0k, detached.
[58.345s][warning][os,thread] Failed to start thread - pthread_create failed (EAGAIN) for attributes: stacksize: 1024k, guardsize: 0k, detached.
runtime: failed to create new OS thread (have 10 already; errno=11)
runtime: may need to increase max user processes (ulimit -u)
fatal error: newosproc
runtime stack:
runtime.throw(0x54c57b, 0x9)
	/opt/hostedtoolcache/go/1.13.14/x64/src/runtime/panic.go:774 +0x72
runtime.newosproc(0xc0000c4700)
	/opt/hostedtoolcache/go/1.13.14/x64/src/runtime/os_linux.go:153 +0x1ba
runtime.newm1(0xc0000c4700)
	/opt/hostedtoolcache/go/1.13.14/x64/src/runtime/proc.go:1853 +0xdc
runtime.newm(0x5555f8, 0xc00002af00)
	/opt/hostedtoolcache/go/1.13.14/x64/src/runtime/proc.go:1832 +0x8f
runtime.startm(0x0, 0xc000026501)
	/opt/hostedtoolcache/go/1.13.14/x64/src/runtime/proc.go:1975 +0x12a
runtime.wakep(...)
	/opt/hostedtoolcache/go/1.13.14/x64/src/runtime/proc.go:2056
runtime.resetspinning()
	/opt/hostedtoolcache/go/1.13.14/x64/src/runtime/proc.go:2436 +0x93
runtime.schedule()
	/opt/hostedtoolcache/go/1.13.14/x64/src/runtime/proc.go:2537 +0x2a0
runtime.park_m(0xc000174000)
	/opt/hostedtoolcache/go/1.13.14/x64/src/runtime/proc.go:2616 +0x9d
runtime.mcall(0x0)
	/opt/hostedtoolcache/go/1.13.14/x64/src/runtime/asm_amd64.s:318 +0x5b
goroutine 1 [syscall]:
syscall.Syscall6(0xf7, 0x1, 0x1db8, 0xc0000a9ce8, 0x1000004, 0x0, 0x0, 0x4056cb, 0xc000014300, 0x0)
	/opt/hostedtoolcache/go/1.13.14/x64/src/syscall/asm_linux_amd64.s:44 +0x5
os.(*Process).blockUntilWaitable(0xc000018810, 0x4df943, 0x644c20, 0x571ea8)
	/opt/hostedtoolcache/go/1.13.14/x64/src/os/wait_waitid.go:31 +0x98
os.(*Process).wait(0xc000018810, 0x10, 0xc000000300, 0x4fe795)
	/opt/hostedtoolcache/go/1.13.14/x64/src/os/exec_unix.go:22 +0x39
os.(*Process).Wait(...)
	/opt/hostedtoolcache/go/1.13.14/x64/src/os/exec.go:125
os/exec.(*Cmd).Wait(0xc000112000, 0x555290, 0xc000112000)
	/opt/hostedtoolcache/go/1.13.14/x64/src/os/exec/exec.go:506 +0x60
main.startJava()
	/home/runner/work/gauge-java/gauge-java/gauge-java.go:240 +0x2b6
main.main()
	/home/runner/work/gauge-java/gauge-java/gauge-java.go:93 +0x2a
goroutine 6 [syscall]:
os/signal.signal_recv(0x0)
	/opt/hostedtoolcache/go/1.13.14/x64/src/runtime/sigqueue.go:147 +0x9c
os/signal.loop()
	/opt/hostedtoolcache/go/1.13.14/x64/src/os/signal/signal_unix.go:23 +0x22
created by os/signal.init.0
	/opt/hostedtoolcache/go/1.13.14/x64/src/os/signal/signal_unix.go:29 +0x41
goroutine 7 [IO wait]:
internal/poll.runtime_pollWait(0x7faa5f5a9f68, 0x72, 0xffffffffffffffff)
	/opt/hostedtoolcache/go/1.13.14/x64/src/runtime/netpoll.go:184 +0x55
internal/poll.(*pollDesc).wait(0xc0000ba1f8, 0x72, 0x8001, 0x8000, 0xffffffffffffffff)
	/opt/hostedtoolcache/go/1.13.14/x64/src/internal/poll/fd_poll_runtime.go:87 +0x45
internal/poll.(*pollDesc).waitRead(...)
	/opt/hostedtoolcache/go/1.13.14/x64/src/internal/poll/fd_poll_runtime.go:92
internal/poll.(*FD).Read(0xc0000ba1e0, 0xc00012a000, 0x8000, 0x8000, 0x0, 0x0, 0x0)
	/opt/hostedtoolcache/go/1.13.14/x64/src/internal/poll/fd_unix.go:169 +0x1cf
os.(*File).read(...)
	/opt/hostedtoolcache/go/1.13.14/x64/src/os/file_unix.go:259
os.(*File).Read(0xc00000e070, 0xc00012a000, 0x8000, 0x8000, 0x91, 0x0, 0x0)
	/opt/hostedtoolcache/go/1.13.14/x64/src/os/file.go:116 +0x71
io.copyBuffer(0x574ce0, 0xc00000c4c0, 0x574d20, 0xc00000e070, 0xc00012a000, 0x8000, 0x8000, 0x0, 0x0, 0x0)
	/opt/hostedtoolcache/go/1.13.14/x64/src/io/io.go:402 +0x122
io.Copy(...)
	/opt/hostedtoolcache/go/1.13.14/x64/src/io/io.go:364
os/exec.(*Cmd).writerDescriptor.func1(0x0, 0x0)
	/opt/hostedtoolcache/go/1.13.14/x64/src/os/exec/exec.go:312 +0x63
os/exec.(*Cmd).Start.func1(0xc000112000, 0xc00000c540)
	/opt/hostedtoolcache/go/1.13.14/x64/src/os/exec/exec.go:440 +0x27
created by os/exec.(*Cmd).Start
	/opt/hostedtoolcache/go/1.13.14/x64/src/os/exec/exec.go:439 +0x632
goroutine 8 [IO wait]:
internal/poll.runtime_pollWait(0x7faa5f5a9dc8, 0x72, 0xffffffffffffffff)
	/opt/hostedtoolcache/go/1.13.14/x64/src/runtime/netpoll.go:184 +0x55
internal/poll.(*pollDesc).wait(0xc0000ba2b8, 0x72, 0x8001, 0x8000, 0xffffffffffffffff)
	/opt/hostedtoolcache/go/1.13.14/x64/src/internal/poll/fd_poll_runtime.go:87 +0x45
internal/poll.(*pollDesc).waitRead(...)
	/opt/hostedtoolcache/go/1.13.14/x64/src/internal/poll/fd_poll_runtime.go:92
internal/poll.(*FD).Read(0xc0000ba2a0, 0xc000132000, 0x8000, 0x8000, 0x0, 0x0, 0x0)
	/opt/hostedtoolcache/go/1.13.14/x64/src/internal/poll/fd_unix.go:169 +0x1cf
os.(*File).read(...)
	/opt/hostedtoolcache/go/1.13.14/x64/src/os/file_unix.go:259
os.(*File).Read(0xc00000e088, 0xc000132000, 0x8000, 0x8000, 0x6f, 0x0, 0x0)
	/opt/hostedtoolcache/go/1.13.14/x64/src/os/file.go:116 +0x71
io.copyBuffer(0x574ce0, 0xc00000c500, 0x574d20, 0xc00000e088, 0xc000132000, 0x8000, 0x8000, 0x0, 0x0, 0x0)
	/opt/hostedtoolcache/go/1.13.14/x64/src/io/io.go:402 +0x122
io.Copy(...)
	/opt/hostedtoolcache/go/1.13.14/x64/src/io/io.go:364
os/exec.(*Cmd).writerDescriptor.func1(0x0, 0x0)
	/opt/hostedtoolcache/go/1.13.14/x64/src/os/exec/exec.go:312 +0x63
os/exec.(*Cmd).Start.func1(0xc000112000, 0xc00000c5a0)
	/opt/hostedtoolcache/go/1.13.14/x64/src/os/exec/exec.go:440 +0x27
created by os/exec.(*Cmd).Start
	/opt/hostedtoolcache/go/1.13.14/x64/src/os/exec/exec.go:439 +0x632
goroutine 17 [chan receive]:
main.listenForKillSignal.func1(0xc0000ba3c0, 0xc000112000)
	/home/runner/work/gauge-java/gauge-java/gauge-java.go:251 +0x38
created by main.listenForKillSignal
	/home/runner/work/gauge-java/gauge-java/gauge-java.go:250 +0xb5
goroutine 18 [sleep]:
runtime.goparkunlock(...)
	/opt/hostedtoolcache/go/1.13.14/x64/src/runtime/proc.go:310
time.Sleep(0x5f5e100)
	/opt/hostedtoolcache/go/1.13.14/x64/src/runtime/time.go:105 +0x157
main.killIfGaugeIsDead(0xc000112000)
	/home/runner/work/gauge-java/gauge-java/gauge-java.go:269 +0x59
created by main.startJava
	/home/runner/work/gauge-java/gauge-java/gauge-java.go:238 +0x2a8
Error occurred while waiting for runner process to finish.


Your Environment Information -----------
	linux, 1.1.1, 6e0d83f
	html-report (4.0.12), java (0.7.9), json-report (0.3.1), spectacle (0.1.4), xml-report (0.2.3)
........................

**Comment by cyberjaime45** [comment]
Gauge version: 1.1.1
Commit Hash: 6e0d83f

Plugins
-------
flash (0.0.2)
html-report (4.0.12)
java (0.7.9)
json-report (0.3.3)
screenshot (0.0.1)

Here's my findings:

You will get "Failed to start gauge API: Timed out connecting to java"  with enable_multithreading=false if you're trying to run 5+ in parallel.

The more powerful you're machine,  the more you can run in parallel.

For example with a Linux machine 
16GB of RAM
4 Cores
You can run in parallel  between 8 to 10 (enable_multithreading=false)

In order to run more in parallel, up to 30 in parallel. I needed to enable -> enable_multithreading=true

When enabling to true, I can easily run my script with a lot of nodes (in my case up to 30 nodes) using the multi thread feature.

However,  when enabling( enable_multithreading=true), you have to make sure your code is thread safe. Otherwise you'll get failures and weird issues.
Thanks to @sriv  for the help.

Because I'm using Saucelabs, I had to change my project to be thread safe.  
https://github.com/saucelabs-training/demo-java/blob/master/parallel-testing/cross-browser-and-platform-config/src/test/java/tests/BaseWebDriverTest.java (This will work for any project.)

I made quite a lot of changes to my script to be thread safe. So far things are looking pretty good.

**Comment by sriv** [comment]
Thanks @cyberjaime45 for posting back your findings. Very useful recommendations.

For those who are still facing this issue:

- if you are running your tests in parallel using `gauge-java` - note that the default behaviour of gauge (as of date) is to spin one runner process per stream. This is important, since this would mean that the resources consumed is linearly proportional to the number of streams and this could easily hit the peak when the stream value is high.

- if you must run tests in parallel, consider running it in multithreaded mode, i.e. set `enable_multithreading=true`.
  - This requires that your code to be Thread aware and Thread safe. The tips mentioned by @cyberjaime45 are very apt.
  - you may also choose to run a static code analysis to detect obvious spots that can lead to concurrency issues.

With all this in place, if you still face this issue, we would need to investigate it deeper, and at this point the best way for a quick resolution would be for us (the gauge maintainers) to get hold of a sample project (or specific steps) to see the issue ourselves.

**Comment by haroon-sheikh** [comment]
@sriv what about these null pointers when multithreading=true 

https://github.com/getgauge/gauge-java/issues/455#issuecomment-662510372
https://github.com/getgauge/gauge-java/issues/455#issuecomment-662590448

Shall we open a separate issue for this? But i've only seen this happening in 0.7.9.

**Comment by sriv** [comment]
Yes, let's open a separate issue for that. It helps to keep the issue focussed, and also helps raise smaller fixes.

**Comment by coding-lama** [comment]
Thanks guys, I was able to finally solve the issue with the information provided here.
